subsys/bluetooth/mesh/adv_multi.c

```diff
@@ -0,0 +1,452 @@
+
+static void send_local_adv(struct k_work *work);
+static void send_relay_adv(struct k_work *work);
+
+#if defined(CONFIG_BT_MESH_GATT_ADV)
+static void send_gatt_adv(struct k_work *work);
+#endif /* CONFIG_BT_MESH_GATT_ADV */
+
+static struct ext_adv {
+	ATOMIC_DEFINE(flags, ADV_FLAGS_NUM);
+	struct bt_le_ext_adv *instance;
+	struct net_buf *buf;
+	uint64_t timestamp;
+	struct k_work_delayable work;
+	struct k_fifo *fifo;
+} adv[BT_MESH_ADV_COUNT] = {
+	[0] = {
+		.work = Z_WORK_DELAYABLE_INITIALIZER(send_local_adv),
+	},
+#if RELAY_ADV_COUNT
+	[1 ... RELAY_ADV_COUNT] = {
+		.work = Z_WORK_DELAYABLE_INITIALIZER(send_relay_adv),
+	},
+#endif /* RELAY_ADV_COUNT */
+#if defined(CONFIG_BT_MESH_GATT_ADV)
+	[BT_MESH_ADV_COUNT - 1] = {
+		.flags = { BIT(ADV_FLAG_PROXY) },
+		.work = Z_WORK_DELAYABLE_INITIALIZER(send_gatt_adv),
+	},
+#endif /* CONFIG_BT_MESH_GATT_ADV */
+};
+
+static int adv_start(struct ext_adv *adv_ins,
+		     const struct bt_le_adv_param *param,
+		     struct bt_le_ext_adv_start_param *start,
+		     const struct bt_data *ad, size_t ad_len,
+		     const struct bt_data *sd, size_t sd_len)
+{
+	int err;
+
+	if (!adv_ins->instance) {
+		BT_ERR("Mesh advertiser not enabled");
+		return -ENODEV;
+	}
+
+	if (atomic_test_and_set_bit(adv_ins->flags, ADV_FLAG_ACTIVE)) {
+		BT_ERR("Advertiser is busy");
+		return -EBUSY;
+	}
+
+	err = bt_le_ext_adv_update_param(adv_ins->instance, param);
+	if (err) {
+		BT_ERR("Failed updating adv params: %d", err);
+		return err;
+	}
+
+	err = bt_le_ext_adv_set_data(adv_ins->instance, ad, ad_len, sd, sd_len);
+	if (err) {
+		BT_ERR("Failed setting adv data: %d", err);
+		atomic_clear_bit(adv_ins->flags, ADV_FLAG_ACTIVE);
+		return err;
+	}
+
+	adv_ins->timestamp = k_uptime_get();
+
+	err = bt_le_ext_adv_start(adv_ins->instance, start);
+	if (err) {
+		BT_ERR("Advertising failed: err %d", err);
+		atomic_clear_bit(adv_ins->flags, ADV_FLAG_ACTIVE);
+	}
+
+	return err;
+}
+
+static int buf_send(struct ext_adv *adv_ins, struct net_buf *buf)
+{
+	struct bt_le_ext_adv_start_param start = {
+		.num_events =
+			BT_MESH_TRANSMIT_COUNT(BT_MESH_ADV(buf)->xmit) + 1,
+	};
+	uint16_t duration, adv_int;
+	struct bt_data ad;
+	int err;
+
+	adv_int = MAX(ADV_INT_FAST_MS,
+		      BT_MESH_TRANSMIT_INT(BT_MESH_ADV(buf)->xmit));
+	/* Upper boundary estimate: */
+	duration = start.num_events * (adv_int + 10);
+
+	BT_DBG("type %u len %u: %s", BT_MESH_ADV(buf)->type,
+	       buf->len, bt_hex(buf->data, buf->len));
+	BT_DBG("count %u interval %ums duration %ums",
+	       BT_MESH_TRANSMIT_COUNT(BT_MESH_ADV(buf)->xmit) + 1, adv_int,
+	       duration);
+
+	ad.type = bt_mesh_adv_type[BT_MESH_ADV(buf)->type];
+	ad.data_len = buf->len;
+	ad.data = buf->data;
+
+	/* Only update advertising parameters if they're different */
+	if (adv_param.interval_min != BT_MESH_ADV_SCAN_UNIT(adv_int)) {
+		adv_param.interval_min = BT_MESH_ADV_SCAN_UNIT(adv_int);
+		adv_param.interval_max = adv_param.interval_min;
+	}
```

Comment:
IMO, it'd be safer to keep `adv_param` on stack now that it's used in so many parallel advertisers. Perhaps the previously configured interval could be stored inside the `ext_adv` struct, so we can keep the old behavior of not updating advertising parameters when the interval is unchanged?